File: 2017/haskell/src/Day05.hs
{-# LANGUAGE BangPatterns #-}



-- module

module Day05 (day05_1, day05_2) where



-- import

import           Control.Monad.ST
import           Data.Vector.Unboxed         (fromList, thaw)
import           Data.Vector.Unboxed.Mutable as V

import           Common



-- solution

day05_1 :: Solution
day05_1 input = show $ runST $ jump update 0 0 =<< toArray (readInt <$> lines input)
  where update d = d + 1

day05_2 :: Solution
day05_2 input = show $ runST $ jump update 0 0 =<< toArray (readInt <$> lines input)
  where update d = if d > 2 then d - 1 else d + 1



-- helpers

type IntArray s = MVector s Int

toArray :: [Int] -> ST s (IntArray s)
toArray = thaw . fromList

jump :: (Int -> Int) -> Int -> Int -> IntArray s -> ST s Int
jump update !jumps !inst steps =
  if inst < 0 || inst >= V.length steps
  then return jumps
  else do
    d <- unsafeRead steps inst
    unsafeWrite steps inst $ update d
    jump update (jumps + 1) (inst + d) steps

File: 2017/rust/src/day_05.rs
// imports

use common::parse_int;



// solutions

pub fn part1(input: &str) -> String {
    let mut steps: Vec<i32> = input.split_whitespace().map(parse_int).collect();
    let res = jumps(&mut steps, |s| s + 1);
    return format!("{}", res);
}

pub fn part2(input: &str) -> String {
    let mut steps: Vec<i32> = input.split_whitespace().map(parse_int).collect();
    let res = jumps(&mut steps, |s| if s > 2 { s-1 } else { s+1 });
    return format!("{}", res);
}



// helpers

fn jumps(steps: &mut Vec<i32>, update: fn(i32) -> i32) -> i32 {
    let mut current: i32 = 0;
    let mut jumps: i32 = 0;

    while current >= 0 && current < (steps.len() as i32) {
        let jump: &mut i32 = steps.get_mut(current as usize).unwrap();
        current += *jump;
        jumps += 1;
        *jump = update(*jump);
    }

    return jumps;
}
